Sources/Datadog/Core/Persistence/Files/File.swift

```diff
@@ -42,27 +43,24 @@ internal struct File: WritableFile, ReadableFile {
     }
 
     /// Appends given data at the end of this file.
-    func append(transaction: ((Data) -> Void) -> Void) throws {
+    func append(transaction: ((Data) throws -> Void) throws -> Void) throws {
         let fileHandle = try FileHandle(forWritingTo: url)
         defer { fileHandle.closeFile() }
-        fileHandle.seekToEndOfFile()
+
+        try objcExceptionHandler.rethrowToSwift {
+            fileHandle.seekToEndOfFile()
+        }
 
         // Writes given data at the end of the file.
-        func appendData(_ data: Data) {
-            /*
-             Apple documentation https://developer.apple.com/documentation/foundation/filehandle/1410936-write says
-             that `fileHandle.write()` raises an exception if no free space is left on the file system,
-             or if any other writing error occurs. Those are unchecked exceptions impossible to handle in Swift.
-
-             It was already escalated to Apple in Swift open source project discussion:
-             https://forums.swift.org/t/pitch-replacement-for-filehandle/5177
-
-             Until better replacement is provided by Apple, the SDK sticks to this API. To mitigate the risk of
-             crashing client applications, other precautions are implemented carefuly.
-             */
-            fileHandle.write(data)
+        func appendData(_ data: Data) throws {
+            try objcExceptionHandler.rethrowToSwift {
+                fileHandle.write(data)
+            }
+        }
+
+        try transaction { chunkOfData in
+            try appendData(chunkOfData)
         }
-        transaction(appendData)
     }
```

Comment:
how does `rethrowToSwift` work when a Swift error is thrown inside `tryBlock`?
for example, would that work?

```swift
try objcExceptionHandler.rethrowToSwift {
  fileHandle.seekToEndOfFile()
  try transaction { chunk in
    fileHandle.write(chunk)
  }
}
```

Reply:
This will not compile, as the `tryBlock {}` signature doesn't include `throws`:
```swift
// swift
override func rethrowToSwift(tryBlock: @escaping () -> Void) throws {

// objective-c
- (BOOL)catchException:(void(^)(void))tryBlock error:(__autoreleasing NSError **)error
    NS_SWIFT_NAME(rethrowToSwift(tryBlock:));
```

That's why we're not able to wrap everything into single `.rethrowToSwift {}` block and I had to isolate objc-throwing parts. Which is IMO good, as we clearly see which APIs might raise an `NSException` and there are not many of them.

Reply:
👍 okay, makes sense
✅ resolved :)